### PR TITLE
Add dev support fix for #81

### PR DESCRIPTION
**Pull request by niklongstone**
This Fixes puli/issues#81
* Add $dev to Packagemanager and Packagemanagerimpl with default = false
* Add Read and Write dev feature in PackageJsonReader and PackageJasonWriter
* Add Package::DEV and refactor Package::match()

### CODE REVIEW

File: src/Api/Package/Package.php [hunk 1]
```diff
@@ -204,6 +218,16 @@ public function isNotFound()
     }
 
     /**
+     * Returns whether the package is dev or not.
+     *
+     * @return bool Returns `true` if the dev state is present.
+     */
+    public function isDev()
```
Comment (webmozart) [inline src/Api/Package/Package.php]: You don't need to add this accessor or a `$dev` property to the package. It's easy enough to get the install info.

File: src/Api/Package/Package.php [hunk 2]
```diff
@@ -232,6 +256,7 @@ public function match(Expression $expr)
             self::INSTALL_PATH => $this->installPath,
             self::STATE => $this->state,
             self::INSTALLER => $this->installInfo ? $this->installInfo->getInstallerName() : null,
+            self::DEV => $this->dev,
```
Comment (webmozart) [inline src/Api/Package/Package.php]: You can do this in the same way as for INSTALLER.

File: src/Package/PackageJsonReader.php [hunk 3]
```diff
@@ -147,7 +147,12 @@ private function populateRootConfig(RootPackageFile $packageFile, \stdClass $jso
 
         if (isset($jsonData->packages)) {
             foreach ($jsonData->packages as $packageName => $packageData) {
-                $installInfo = new InstallInfo($packageName, $packageData->{'install-path'});
+                $isDev = false;
+                if (isset( $packageData->{'dev'})) {
```
Comment (webmozart) [inline src/Package/PackageJsonReader.php]: You can skip the braces "{". They are only required if the property name contains special characters like "-". Also there's an extra space.

File: tests/Api/Package/PackageTest.php [hunk 4]
```diff
@@ -93,27 +94,41 @@ public function testCreatePackageWithoutPackageFileNorInstallInfo()
...
+
+        $this->assertSame(true, $package->isDev());
+    }
+*/
     public function testMatch()
     {
-        $packageFile = new PackageFile('vendor/name');
-        $package = new Package($packageFile, __DIR__);
+       $packageFile = new PackageFile('vendor/name');
+       $package = new Package($packageFile, __DIR__);
```
Comment (webmozart) [inline tests/Api/Package/PackageTest.php]: You removed a space everywhere here.

File: src/Package/PackageJsonWriter.php [hunk 5]
```diff
@@ -227,6 +227,10 @@ private function addRootConfig(array &$jsonData, RootPackageFile $packageFile)
                     sort($installData->{'disabled-bindings'});
                 }
 
+                if ($installInfo->isDev()) {
+                    $installData->{'dev'} = $installInfo->isDev();
```
Comment (webmozart) [inline src/Package/PackageJsonWriter.php]: You can skip the braces here too.

File: src/Api/Package/PackageManager.php [hunk 6]
```diff
@@ -38,12 +38,13 @@ public function getEnvironment();
...
      *                                   puli.json.
      * @param string      $installerName The name of the installer.
+     * @param bool        $isDev         The package dev parameter.gis
      *
      * @throws InvalidConfigException If the package is not configured correctly.
      * @throws NameConflictException If the package has the same name as another
      *                               loaded package.
      */
-    public function installPackage($installPath, $name = null, $installerName = InstallInfo::DEFAULT_INSTALLER_NAME);
+    public function installPackage($installPath, $name = null, $installerName = InstallInfo::DEFAULT_INSTALLER_NAME, $isDev = false);
```
Comment (webmozart) [inline src/Api/Package/PackageManager.php]: Could you name the property `$dev` here and everywhere else? Usually in Puli, boolean properties don't have the prefix "is".

File: tests/Api/Package/PackageTest.php [hunk 7]
```diff
@@ -26,6 +26,7 @@
  */
 class PackageTest extends PHPUnit_Framework_TestCase
 {
+    /*
```
Comment (webmozart) [inline tests/Api/Package/PackageTest.php]: Attention, you commented the following tests out.

**Comment by webmozart** [comment]
Thanks for working on this @niklongstone! :) See my inline comments for what needs to be changed.

**Comment by niklongstone** [comment]
Updated :wrench:

**Comment by webmozart** [comment]
Thanks :) I merged this now and improved a few more glitches in https://github.com/puli/manager/commit/d63f76e062f8df68d648db17b000737c5601456e.

**Pull request closed without merging**